openvpn/middlewares/server/auth/middleware.go

```diff
@@ -37,143 +60,104 @@ func (m *middleware) Stop(commandWriter management.Connection) error {
+	switch eventType {
+	case connect, reauth:
+		ID, key, err := parseIDAndKey(eventData)
+		if err != nil {
+			return true, err
+		}
+		m.startOfEvent(eventType, ID, key)
+	case env:
+		if strings.ToLower(eventData) == "end" {
+			m.endOfEvent()
+			return true, nil
 		}
-		m.lastPassword = match[1]
-		return false, true, nil
-	}
-
-	return true, false, nil
-}
-
-func (m *middleware) checkUsername(line string) (cont bool, consumed bool, err error) {
-
-	rule, err := regexp.Compile("^>CLIENT:ENV,username=(.*)$")
-	if err != nil {
-		return false, false, err
-	}
 
-	match := rule.FindStringSubmatch(line)
-	if len(match) > 0 {
-		if m.clientID < 0 {
-			return false, false, fmt.Errorf("wrong auth state, no client id")
+		key, val, err := parseEnvVar(eventData)
+		if err != nil {
+			return true, err
 		}
-		m.lastUsername = match[1]
-		return false, true, nil
+		m.addEnvVar(key, val)
+	case established, disconnect:
+		ID, err := parseID(eventData)
+		if err != nil {
+			return true, err
+		}
+		m.startOfEvent(eventType, ID, undefined)
+	case address:
+		log.Info("Address for client: ", eventData)
+	default:
+		log.Error("Undefined user notification event: ", eventType, eventData)
+		log.Error("Original line was: ", line)
 	}
-
-	return true, false, nil
+	return true, nil
 }
 
-func (m *middleware) checkEnvEnd(line string) (cont bool, consumed bool, err error) {
-
-	rule, err := regexp.Compile("^>CLIENT:ENV,END$")
-	if err != nil {
-		return false, false, err
-	}
-
-	match := rule.FindStringSubmatch(line)
-	if len(match) > 0 {
-		return false, true, nil
-	}
-
-	return true, false, nil
+func (m *middleware) startOfEvent(eventType clientEventType, clientID int, keyID int) {
+	m.currentEvent.eventType = eventType
+	m.currentEvent.clientID = clientID
+	m.currentEvent.clientKey = keyID
 }
 
-func (m *middleware) ConsumeLine(line string) (consumed bool, err error) {
-	if cont, consumed, err := m.checkReAuth(line); !cont {
-		return consumed, err
-	}
-
-	if cont, consumed, err := m.checkConnect(line); !cont {
-		return consumed, err
-	}
+func (m *middleware) addEnvVar(key string, val string) {
+	m.currentEvent.env[key] = val
+}
 
-	if cont, consumed, err := m.checkUsername(line); !cont {
-		return consumed, err
-	}
+func (m *middleware) endOfEvent() {
+	m.handleClientEvent(m.currentEvent)
+	m.reset()
+}
 
-	if cont, consumed, err := m.checkPassword(line); !cont {
-		return consumed, err
-	}
+func (m *middleware) reset() {
+	m.currentEvent = undefinedEvent
+}
 
-	if cont, consumed, err := m.checkEnvEnd(line); !cont {
-		if consumed {
-			return m.authenticateClient()
+func (m *middleware) handleClientEvent(event clientEvent) {
+	switch event.eventType {
+	case connect, reauth:
+		username := event.env["username"]
+		password := event.env["password"]
+		err := m.authenticateClient(event.clientID, event.clientKey, username, password)
+		if err != nil {
+			log.Critical("Unable to authenticate client. Error: ", err)
```

Comment:
I woudn't say its critical. Critical usually means that application cannot proceed. In this case just 1 client..